### PR TITLE
KAFKA-13511: Add support for different unix precisions in TimestampConverter SMT

### PR DESCRIPTION
**Pull request by jchanaud**
Currently, the SMT TimestampConverter can convert Timestamp from either source String, Long or Date into target String, Long or Date.

The problem is that Long source or target is required to be epoch in milliseconds.

In many cases, epoch is represented with different precisions. This leads to several Jira tickets :

[KAFKA-12364](https://issues.apache.org/jira/browse/KAFKA-12364): add support for date from int32 to timestampconverter
[KAFKA-10561](https://issues.apache.org/jira/browse/KAFKA-10561): Support microseconds precision for Timestamps
I propose to add a new config to TimestampConverter called "epoch.precision" which defaults to "millis" so as to not impact existing code, and allows for more precisions : seconds, millis, micros.
````json
"transforms": "TimestampConverter",
"transforms.TimestampConverter.type": "org.apache.kafka.connect.transforms.TimestampConverter$Value",
"transforms.TimestampConverter.field": "event_date",
"transforms.TimestampConverter.epoch.precision": "micros",
"transforms.TimestampConverter.target.type": "Timestamp"
````
Exactly like "format" field which is used as input when the source in String and output when the target.type is string, this new field would be used as input when the field is Long, and as output when the target.type is "unix"

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by jchanaud** [comment]
@mimaison @rhauch before considering going further with this PR, any chance I could to get your opinion on this subject ?

The main idea behind this PR is that Avro logical types are not fully supported in Kafka Connect but even if they were, it would not help schema-less messages that need to convert epoch with a precision different than ms.

For this reason, I consider this would be an interesting addition.
Thanks for your help.

**Comment by mimaison** [comment]
Hi @twobeeb, thanks for the PR, this looks like a useful addition. However as this is adding a new configuration, this change requires a KIP before we can accept it.
Take a look at https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals. This page details the KIP process. Let me know if you have any questions.

**Comment by jchanaud** [comment]
Hi @mimaison, 
I haven't received any update on the [KIP-808](https://lists.apache.org/thread/t51lmxjdt3k4y990s2c378529lwtt0q0) which I created as per your suggestion.

Could you be of any help ?

**Comment by mimaison** [comment]
Sorry @twobeeb, I'll try to take a look this afternoon. Thanks

**Comment by jchanaud** [comment]
Hi everyone,
Since [KIP-808](https://cwiki.apache.org/confluence/display/KAFKA/KIP-808%3A+Add+support+for+different+unix+precisions+in+TimestampConverter+SMT) is now adopted, could you review and comment this PR ?
Thanks

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java [hunk 1]
```diff
@@ -65,14 +66,22 @@
...
             .define(FIELD_CONFIG, ConfigDef.Type.STRING, FIELD_DEFAULT, ConfigDef.Importance.HIGH,
                     "The field containing the timestamp, or empty if the entire value is a timestamp")
             .define(TARGET_TYPE_CONFIG, ConfigDef.Type.STRING, ConfigDef.Importance.HIGH,
                     "The desired timestamp representation: string, unix, Date, Time, or Timestamp")
             .define(FORMAT_CONFIG, ConfigDef.Type.STRING, FORMAT_DEFAULT, ConfigDef.Importance.MEDIUM,
                     "A SimpleDateFormat-compatible format for the timestamp. Used to generate the output when type=string "
-                            + "or used to parse the input if the input is a string.");
+                            + "or used to parse the input if the input is a string.")
+            .define(UNIX_PRECISION_CONFIG, ConfigDef.Type.STRING, UNIX_PRECISION_DEFAULT, ConfigDef.Importance.LOW,
+                    "The desired unix precision for the timestamp. Used to generate the output when type=unix " +
```
Comment (tombentley) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]:
```suggestion
                    "The desired Unix precision for the timestamp. Used to generate the output when type=unix " +
```

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java [hunk 2]
```diff
@@ -65,14 +66,22 @@
...
+                            "or used to parse the input if the input is a Long." +
+                            "Note: This SMT will cause precision loss during conversions from and to values with sub-milliseconds components.");
```
Comment (tombentley) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]:
```suggestion
                            "Note: This SMT will cause precision loss during conversions from, and to, values with sub-millisecond components.");
```

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java [hunk 3]
```diff
@@ -65,14 +66,22 @@
...
     public static final ConfigDef CONFIG_DEF = new ConfigDef()
             .define(FIELD_CONFIG, ConfigDef.Type.STRING, FIELD_DEFAULT, ConfigDef.Importance.HIGH,
                     "The field containing the timestamp, or empty if the entire value is a timestamp")
             .define(TARGET_TYPE_CONFIG, ConfigDef.Type.STRING, ConfigDef.Importance.HIGH,
                     "The desired timestamp representation: string, unix, Date, Time, or Timestamp")
             .define(FORMAT_CONFIG, ConfigDef.Type.STRING, FORMAT_DEFAULT, ConfigDef.Importance.MEDIUM,
                     "A SimpleDateFormat-compatible format for the timestamp. Used to generate the output when type=string "
-                            + "or used to parse the input if the input is a string.");
+                            + "or used to parse the input if the input is a string.")
+            .define(UNIX_PRECISION_CONFIG, ConfigDef.Type.STRING, UNIX_PRECISION_DEFAULT, ConfigDef.Importance.LOW,
```
Comment (tombentley) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]: There's an overload of `define()` that takes a validator, which should mean you don't need to validate the value yourself and also ensure the valid values are documented;
```suggestion
            .define(UNIX_PRECISION_CONFIG, ConfigDef.Type.STRING, UNIX_PRECISION_DEFAULT,
                    ConfigDef.ValidString.in(
                            UNIX_PRECISION_NANOS, UNIX_PRECISION_MICROS,
                            UNIX_PRECISION_MILLIS, UNIX_PRECISION_SECONDS), 
                    ConfigDef.Importance.LOW,
```

**Review by tombentley — COMMENTED** [review]
Thanks for the PR!

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java (hunk 3 above)
Comment (jchanaud) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]: Agreed, I've been misled by existing validation code which spans across 2 fields, making it impossible to use Validator. I'll update accordingly.

**Comment by jchanaud** [comment]
@tombentley 
Thanks for your inputs. 
Regarding your last comment which I fixed, the same case could be made for `target.type`.
https://github.com/apache/kafka/blob/trunk/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java#L254-L257
Should we address this ? If so, this PR or new one ?

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java [hunk 4]
```diff
@@ -83,12 +78,36 @@
     private static final String TYPE_TIMESTAMP = "Timestamp";
     private static final Set<String> VALID_TYPES = new HashSet<>(Arrays.asList(TYPE_STRING, TYPE_UNIX, TYPE_DATE, TYPE_TIME, TYPE_TIMESTAMP));
 
+    private static final String UNIX_PRECISION_MILLIS = "milliseconds";
+    private static final String UNIX_PRECISION_MICROS = "microseconds";
+    private static final String UNIX_PRECISION_NANOS = "nanoseconds";
+    private static final String UNIX_PRECISION_SECONDS = "seconds";
+    private static final Set<String> VALID_UNIX_PRECISIONS = new HashSet<>(
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]: This field is not used, we can remove it.

**Review by mimaison — COMMENTED** [review]
Thanks @twobeeb for the PR. It looks good, I just left a minor comment.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java (hunk 4 above)
Comment (jchanaud) [reply connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]: Thanks, sorry about this

**Comment by mimaison** [comment]
> @tombentley Thanks for your inputs. Regarding your last comment which I fixed, the same case could be made for `target.type`. https://github.com/apache/kafka/blob/trunk/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java#L254-L257 Should we address this ? If so, this PR or new one ?

Yes feel free to address it in this PR if you want.

File: connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java [hunk 5]
```diff
@@ -83,12 +78,36 @@
     private static final String TYPE_TIMESTAMP = "Timestamp";
     private static final Set<String> VALID_TYPES = new HashSet<>(Arrays.asList(TYPE_STRING, TYPE_UNIX, TYPE_DATE, TYPE_TIME, TYPE_TIMESTAMP));
```
Comment (mimaison) [inline connect/transforms/src/main/java/org/apache/kafka/connect/transforms/TimestampConverter.java]: `VALID_TYPES` can now be removed too

**Review by mimaison — APPROVED** [review]
LGTM, thanks for the PR!

**Pull request merged**